File: agenkit-scala/src/main/scala/io/agenkit/patterns/ConversationalAgent.scala
package io.agenkit.patterns

import io.agenkit.adapters.LlmClient
import io.agenkit.core.*
import scala.concurrent.{ExecutionContext, Future}
import java.util.concurrent.atomic.AtomicLong

/** Stateful agent that maintains a conversation history across turns. */
class ConversationalAgent(
  val name: String,
  llm: LlmClient,
  systemPrompt: Option[String] = None,
  maxHistorySize: Int = 100
) extends Agent:
  private val history          = collection.mutable.ListBuffer[Message]()
  private val _processedCount  = new AtomicLong(0)

  def capabilities: List[String] = List("conversation", "context-aware")

  def process(message: Message)(using ExecutionContext): Future[Message] =
    history += message
    val messages = systemPrompt
      .map(p => Message.system(p) :: history.toList)
      .getOrElse(history.toList)
    llm.complete(messages).map { response =>
      _processedCount.incrementAndGet()
      history += response
      if history.size > maxHistorySize * 2 then
        history.remove(0, history.size - maxHistorySize * 2)
      response
    }

  def clearHistory(): Unit         = history.clear()
  def getHistory: List[Message]    = history.toList

  def introspect(): IntrospectionResult =
    IntrospectionResult(
      name = name,
      metadata = Map("historySize" -> history.size, "maxHistorySize" -> maxHistorySize),
      capabilities = capabilities,
      processedMessages = _processedCount.get()
    )

File: agenkit-java/src/main/java/io/agenkit/patterns/ConversationalAgent.java
package io.agenkit.patterns;

import io.agenkit.adapters.LlmClient;
import io.agenkit.core.Agent;
import io.agenkit.core.IntrospectionResult;
import io.agenkit.core.Message;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * An agent that maintains conversation history.
 */
public final class ConversationalAgent implements Agent {

    private static final int DEFAULT_MAX_HISTORY = 10;

    private final String name;
    private final LlmClient llmClient;
    private final String systemPrompt;
    private int maxHistory;
    private final List<Message> history = new ArrayList<>();

    public ConversationalAgent(String name, LlmClient llmClient, String systemPrompt, int maxHistory) {
        this.name = name;
        this.llmClient = llmClient;
        this.systemPrompt = systemPrompt;
        this.maxHistory = maxHistory;
    }

    public ConversationalAgent(String name, LlmClient llmClient) {
        this(name, llmClient, "You are a helpful assistant.", DEFAULT_MAX_HISTORY);
    }

    @Override
    public String getName() { return name; }

    @Override
    public List<String> getCapabilities() {
        return List.of("conversation", "history");
    }

    @Override
    public CompletableFuture<Message> process(Message message) {
        history.add(message);
        List<Message> messages = buildMessages();
        return llmClient.complete(messages).thenApply(response -> {
            history.add(response);
            pruneHistory();
            return response;
        });
    }

    @Override
    public IntrospectionResult introspect() {
        Map<String, Object> state = new HashMap<>();
        state.put("historySize", history.size());
        state.put("maxHistory", maxHistory);
        return new IntrospectionResult(name, getCapabilities(), null, state, null);
    }

    public List<Message> getHistory() { return Collections.unmodifiableList(history); }

    public void clearHistory(boolean keepSystem) {
        history.clear();
    }

    public void setMaxHistory(int maxHistory) {
        this.maxHistory = maxHistory;
        pruneHistory();
    }

    private List<Message> buildMessages() {
        List<Message> messages = new ArrayList<>();
        if (systemPrompt != null && !systemPrompt.isEmpty()) {
            messages.add(Message.of("system", systemPrompt));
        }
        messages.addAll(history);
        return messages;
    }

    private void pruneHistory() {
        while (history.size() > maxHistory * 2) {
            history.remove(0);
            if (!history.isEmpty()) {
                history.remove(0);
            }
        }
    }
}
